components/script/body.rs

```diff
@@ -26,41 +54,105 @@ pub fn consume_body<T: BodyTrait + Reflectable>(object: &T, body_type: BodyType)
-fn run_package_data_algorithm(bytes: Option<Vec<u8>>,
-                              body_type: BodyType, mime_type: Ref<Vec<u8>>) -> Fallible<String> {
+#[allow(unsafe_code)]
+fn run_package_data_algorithm<T: BodyTrait + Reflectable>(object: &T,
+                                                          bytes: Option<Vec<u8>>,
+                                                          body_type: BodyType,
+                                                          mime_type: Ref<Vec<u8>>) -> Fallible<FetchedData> {
+    if let Some(b) = bytes {
+        let cx = object.global().r().get_cx();
+        let mime = &*mime_type;
         match body_type {
-            BodyType::Text => {
-                // return result of running utf-8 decode here
-                // using encoding crate on all bytes instead of
-                // individually processing each token
-                if let Some(bytes) = bytes {
-                    UTF_8.decode(&bytes, DecoderTrap::Replace).unwrap();
-                }
-                return Ok("".to_owned());
-            },
-            _ => Err(Error::Type("Unable to process bytes".to_string()))
+            BodyType::Text => run_text_data_algorithm(b),
+            BodyType::Json => run_json_data_algorithm(cx, b),
+            BodyType::Blob => run_blob_data_algorithm(object.global().r(), b, mime),
+            BodyType::FormData => run_form_data_algorithm(object.global().r(), b, mime),
+            _ => Err(Error::Type("Unable to process body type".to_string()))
         }
+    } else {
+        return Err(Error::Type("Unable to process bytes".to_string()));
+    }
 }
 
-pub enum BodyType {
-    ArrayBuffer,
-    Blob,
-    FormData,
-    Json,
-    Text
+fn run_text_data_algorithm(bytes: Vec<u8>) -> Fallible<FetchedData> {
+    let text = UTF_8.decode(&bytes, DecoderTrap::Replace).unwrap();
+    Ok(FetchedData::Text(text))
+}
+
+#[allow(unsafe_code)]
+fn run_json_data_algorithm(cx: *mut JSContext,
+                           bytes: Vec<u8>) -> Fallible<FetchedData> {
+    let json_text = UTF_8.decode(&bytes, DecoderTrap::Replace).unwrap();
+    let json_text: Vec<u16> = json_text.encode_utf16().collect();
+    rooted!(in(cx) let mut rval = UndefinedValue());
+    unsafe {
+        if !JS_ParseJSON(cx,
+                         json_text.as_ptr(),
+                         json_text.len() as u32,
+                         rval.handle_mut()) {
+            JS_ClearPendingException(cx);
+            // TODO
+            // ... Should this return `NullValue()`?
+            return Err(Error::Type("Failed to parse JSON".to_string()));
+            }
+        Ok(FetchedData::Json(rval.get()))
+    }
+}
+
+fn run_blob_data_algorithm(root: GlobalRef,
+                           bytes: Vec<u8>,
+                           mime: &[u8]) -> Fallible<FetchedData> {
+    let mime_string = if let Ok(s) = String::from_utf8(mime.to_vec()) {
+        s
+    } else {
+        "None".to_string()
```

Comment:
I don't see any text in the specification backing this up. I believe an empty mime type is supposed to be represented as the empty string instead.